### PR TITLE
Fix `test_long_log_msg`

### PR DESCRIPTION
**Pull request by ktbarrett**
From https://github.com/cocotb/cocotb/pull/2924#issuecomment-1128245463.

### CODE REVIEW

**Comment by ktbarrett** [comment]
Apparently Window's version of `vsnprintf` when encountering a potential buffer overflow returns `-1` instead of the number of characters which would have been written, violating the C99 standard. This affects code compiled with MSVC and mingw, apparently is fixed in UCRT, and may or may not work in msys/cygwin. Guess I'll just have to use a dumber algorithm...

**Comment by ktbarrett** [comment]
The new fix is now only slower and dumber on Windows and should continue working like it did previously on other OSes.

File: cocotb/share/lib/gpi_log/gpi_logging.cpp [hunk 1]
```diff
@@ -129,18 +129,31 @@ extern "C" void gpi_native_logger_vlog(const char *name, int level,
...
-        return;
-        // LCOV_EXCL_STOP
-    } else if ((unsigned)n >= log_buff.capacity()) {
+        // On Windows the above call to vsnprintf will return -1 if the buffer
+        // would overflow, rather than the number of bytes that would be written
+        // (as required by the C99 standard). So we try the call again with the
+        // buffer NULL and the size 0, which should return the number of bytes
+        // that would be written, even on Windows...
+        va_list argp_copy_again;
+        va_copy(argp_copy_again, argp);
```
Comment (marlonjames) [inline cocotb/share/lib/gpi_log/gpi_logging.cpp:138]:
```suggestion
        va_copy(argp_copy_again, argp_copy);
```
`argp` is already used at this point.

**Comment by ktbarrett** [comment]
Something is not right... `cat`ing out the error log we get [this](https://github.com/cocotb/cocotb/runs/6464420830?check_suite_focus=true#step:19:2260). The error message can only be `Error constructing log message` followed by `1` or `2`, a colon, then the error code. That's not what we are seeing, so I'm not sure where this is coming from? Maybe I'm just loosing my mind.

**Comment by ktbarrett** [comment]
Ohnonono, we aren't actually testing cocotb on Windows, the build fails and it just pulls down cocotb 1.6.2 from pypi. https://github.com/cocotb/cocotb/runs/6464420830?check_suite_focus=true#step:16:28

**Comment by marlonjames** [comment]
Whatever we do in `gpi_logging.cpp` we need to copy over to `py_gpi_logging.cpp` because that is the actual handler.

**Comment by imphil** [comment]
Maybe a stupid question: can we link against/use UCRT and simply get the C99-conforming vnsprintf() function? (Even msys seems to support UCRT, from my reading of https://www.msys2.org/docs/environments/ -- and we primarily care about MSVC anyways.)

**Comment by imphil** [comment]
Just to be clear: I'd be interested in exploring the use of UCRT. But that's optional, the approach in this PR is clearly sufficient for now and gets us past this failure. Let me know @ktbarrett when you'd like a review to happen.

**Comment by ktbarrett** [comment]
@imphil The mingw compiler test still fails on master, I have no idea if this fix works or not.

**Comment by marlonjames** [comment]
This might have some info: https://mingwpy.github.io/ucrt.html

**Comment by ktbarrett** [comment]
This PR is now good to come in despite mingw compiler not working because we no longer support that compiler on Windows, right?

**Comment by imphil** [comment]
Yeah, let's confirm that by rebasing this PR on top of the latest master.

**Review by marlonjames — CHANGES_REQUESTED** [review]
The test is still disabled in the Makefile.

It's not run in the windows CI run (or any CI run):
https://github.com/cocotb/cocotb/runs/7042194080?check_suite_focus=true#step:17:3117

**Comment by ktbarrett** [comment]
Codecov is all kinds of screwed up on this PR, but the change should be complete.

File: cocotb/share/lib/gpi_log/gpi_logging.cpp [hunk 2]
```diff
@@ -129,18 +129,31 @@ extern "C" void gpi_native_logger_vlog(const char *name, int level,
...
-        fprintf(stderr, "Log message construction failed: (error code) %d\n",
-                n);
-        return;
-        // LCOV_EXCL_STOP
-    } else if ((unsigned)n >= log_buff.capacity()) {
+        // On Windows the above call to vsnprintf will return -1 if the buffer
+        // would overflow, rather than the number of bytes that would be written
+        // (as required by the C99 standard). So we try the call again with the
+        // buffer NULL and the size 0, which should return the number of bytes
+        // that would be written, even on Windows...
```
[the remark this answers is not in the document]
Comment (ktbarrett) [reply cocotb/share/lib/gpi_log/gpi_logging.cpp]: Added a link to the MSDN documentation which has more information.

**Review by marlonjames — APPROVED** [review]

**Pull request merged**